File: automation/scripts/extract_swe_bench_corpus.py
#!/usr/bin/env python3
"""
SWE-bench Corpus Extractor

This script extracts actual repository content for SWE-bench queries, creating a proper 
corpus that matches what the queries are actually looking for. The key insight is that
each benchmark needs its own corpus that aligns with the query domain.

SWE-bench queries are about real GitHub issues/PRs, so they need actual repository code.
"""

import json
import os
import subprocess
import tempfile
import shutil
from pathlib import Path
from collections import Counter, defaultdict
import argparse
import logging

# Setup logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

def get_unique_repositories(swe_bench_data):
    """Extract unique repositories from SWE-bench dataset"""
    repositories = set()
    repo_counts = Counter()
    
    for entry in swe_bench_data:
        repo_name = entry['repository']
        repositories.add(repo_name)
        repo_counts[repo_name] += 1
    
    logger.info(f"Found {len(repositories)} unique repositories across {len(swe_bench_data)} entries")
    logger.info(f"Top 10 repositories: {repo_counts.most_common(10)}")
    
    return repositories, repo_counts

def clone_repository(repo_name, base_commit, clone_dir):
    """Clone a repository at a specific commit"""
    repo_url = f"https://github.com/{repo_name}.git"
    repo_path = clone_dir / repo_name.replace('/', '_')
    
    try:
        # Clone the repository
        subprocess.run(['git', 'clone', repo_url, str(repo_path)], 
                      check=True, capture_output=True)
        
        # Checkout specific commit
        subprocess.run(['git', 'checkout', base_commit], 
                      cwd=repo_path, check=True, capture_output=True)
        
        logger.info(f"Successfully cloned {repo_name} at commit {base_commit[:8]}")
        return repo_path
        
    except subprocess.CalledProcessError as e:
        logger.error(f"Failed to clone {repo_name}: {e}")
        return None

def extract_source_files(repo_path, output_dir, repo_name):
    """Extract relevant source files from cloned repository"""
    if not repo_path or not repo_path.exists():
        return []
    
    # File extensions to include
    source_extensions = {
        '.py', '.js', '.ts', '.jsx', '.tsx', '.java', '.cpp', '.cc', '.cxx', '.c', 
        '.h', '.hpp', '.cs', '.rb', '.go', '.rs', '.php', '.scala', '.kt', '.swift',
        '.m', '.mm', '.sh', '.sql', '.html', '.css', '.scss', '.less', '.vue',
        '.md', '.rst', '.txt', '.yaml', '.yml', '.json', '.xml', '.toml', '.ini'
    }
    
    # Directories to skip
    skip_dirs = {
        '.git', '.svn', '.hg', '__pycache__', 'node_modules', '.tox', 'venv', 
        'env', '.env', 'build', 'dist', '.pytest_cache', '.coverage', 
        'htmlcov', '.idea', '.vscode', 'target', 'bin', 'obj'
    }
    
    extracted_files = []
    file_counter = 0
    
    for root, dirs, files in os.walk(repo_path):
        # Skip unwanted directories
        dirs[:] = [d for d in dirs if d not in skip_dirs]
        
        for file in files:
            file_path = Path(root) / file
            
            # Check if file has a relevant extension
            if file_path.suffix.lower() in source_extensions:
                try:
                    # Read file content
                    with open(file_path, 'r', encoding='utf-8', errors='ignore') as f:
                        content = f.read()
                    
                    # Skip empty files or files that are too large
                    if 10 <= len(content) <= 100000:  # Between 10 chars and 100KB
                        # Create output filename
                        safe_repo_name = repo_name.replace('/', '_')
                        relative_path = file_path.relative_to(repo_path)
                        safe_relative_path = str(relative_path).replace('/', '_').replace('\\', '_')
                        output_filename = f"{safe_repo_name}_{safe_relative_path}"
                        
                        # Write to corpus
                        output_path = output_dir / output_filename
                        with open(output_path, 'w', encoding='utf-8') as f:
                            f.write(content)
                        
                        extracted_files.append(output_filename)
                        file_counter += 1
                        
                        if file_counter % 100 == 0:
                            logger.info(f"Extracted {file_counter} files from {repo_name}")
                
                except (IOError, UnicodeDecodeError):
                    continue  # Skip files that can't be read
    
    logger.info(f"Extracted {len(extracted_files)} files from {repo_name}")
    return extracted_files

def process_repository_sample(repo_name, entries, clone_dir, output_dir, max_commits=3):
    """Process a sample of commits from a repository"""
    # Sort entries by creation date and take latest commits
    entries_sorted = sorted(entries, key=lambda x: x['created_at'], reverse=True)
    sample_entries = entries_sorted[:max_commits]
    
    all_extracted_files = []
    
    for entry in sample_entries:
        base_commit = entry['base_commit']
        
        # Clone repository at this commit
        repo_path = clone_repository(repo_name, base_commit, clone_dir)
        
        if repo_path:
            # Extract source files
            extracted_files = extract_source_files(repo_path, output_dir, repo_name)
            all_extracted_files.extend(extracted_files)
            
            # Clean up cloned repo
            shutil.rmtree(repo_path)
        
        # Don't extract from too many commits of the same repo
        if len(all_extracted_files) > 500:  # Reasonable limit per repository
            break
    
    return all_extracted_files

def create_corpus_directories():
    """Create corpus directory structure"""
    base_dir = Path("benchmark-corpus")
    
    corpus_dirs = {
        'swe-bench': base_dir / "swe-bench",
        'codesearchnet': base_dir / "codesearchnet", 
        'coir': base_dir / "coir",
        'cosqa': base_dir / "cosqa"
    }
    
    # Create directories
    for name, path in corpus_dirs.items():
        path.mkdir(parents=True, exist_ok=True)
        logger.info(f"Created directory: {path}")
    
    return corpus_dirs

def main():
    parser = argparse.ArgumentParser(description='Extract SWE-bench corpus from repository content')
    parser.add_argument('--dataset', default='datasets/swe-bench-verified.json', 
                       help='Path to SWE-bench dataset')
    parser.add_argument('--max-repos', type=int, default=20,
                       help='Maximum number of repositories to process')
    parser.add_argument('--max-commits-per-repo', type=int, default=3,
                       help='Maximum commits to sample per repository')
    parser.add_argument('--dry-run', action='store_true',
                       help='Show what would be done without actually doing it')
    
    args = parser.parse_args()
    
    # Load SWE-bench dataset
    logger.info(f"Loading SWE-bench dataset from {args.dataset}")
    with open(args.dataset, 'r') as f:
        swe_bench_data = json.load(f)
    
    # Get unique repositories
    repositories, repo_counts = get_unique_repositories(swe_bench_data)
    
    if args.dry_run:
        logger.info("DRY RUN - Would process:")
        for repo, count in repo_counts.most_common(args.max_repos):
            logger.info(f"  {repo}: {count} entries")
        return
    
    # Create corpus directories
    corpus_dirs = create_corpus_directories()
    
    # Group entries by repository
    entries_by_repo = defaultdict(list)
    for entry in swe_bench_data:
        entries_by_repo[entry['repository']].append(entry)
    
    # Process top repositories
    with tempfile.TemporaryDirectory() as temp_dir:
        clone_dir = Path(temp_dir)
        
        total_files = 0
        processed_repos = 0
        
        for repo_name, count in repo_counts.most_common(args.max_repos):
            if processed_repos >= args.max_repos:
                break
                
            logger.info(f"Processing repository {repo_name} ({count} entries)")
            
            entries = entries_by_repo[repo_name]
            extracted_files = process_repository_sample(
                repo_name, entries, clone_dir, corpus_dirs['swe-bench'], 
                args.max_commits_per_repo
            )
            
            total_files += len(extracted_files)
            processed_repos += 1
            
            logger.info(f"Progress: {processed_repos}/{args.max_repos} repos, {total_files} total files")
    
    # Write corpus summary
    summary = {
        'total_repositories_processed': processed_repos,
        'total_files_extracted': total_files,
        'corpus_type': 'swe-bench',
        'source_dataset': args.dataset,
        'extraction_params': {
            'max_repos': args.max_repos,
            'max_commits_per_repo': args.max_commits_per_repo
        }
    }
    
    summary_path = corpus_dirs['swe-bench'] / 'corpus_summary.json'
    with open(summary_path, 'w') as f:
        json.dump(summary, f, indent=2)
    
    logger.info(f"Extraction complete! Summary written to {summary_path}")
    logger.info(f"Total: {processed_repos} repositories, {total_files} files")

if __name__ == '__main__':
    main()

File: automation/scripts/extract-swe-bench-corpus.js
#!/usr/bin/env node

/**
 * SWE-Bench Corpus Extractor
 * 
 * This script extracts real repository code from SWE-bench verified dataset
 * and populates the benchmark-corpus directory with actual source files.
 * 
 * Repositories to extract from SWE-bench:
 * - django/django (231 entries) - Web framework
 * - sympy/sympy (75 entries) - Symbolic math library  
 * - sphinx-doc/sphinx (44 entries) - Documentation generator
 * - matplotlib/matplotlib (34 entries) - Plotting library
 * - scikit-learn/scikit-learn (32 entries) - Machine learning
 * - astropy/astropy (22 entries) - Astronomy library
 * - pydata/xarray (22 entries) - Data analysis
 * - pytest-dev/pytest (19 entries) - Testing framework
 * - pylint-dev/pylint (10 entries) - Code linter
 * - psf/requests (8 entries) - HTTP library
 * - mwaskom/seaborn (2 entries) - Statistical visualization
 * - pallets/flask (1 entry) - Web framework
 */

import fs from 'fs';
import path from 'path';
import { execSync } from 'child_process';

const REPOS_DIR = './swe-bench-repos';
const CORPUS_DIR = './benchmark-corpus';
const MAX_FILES_PER_REPO = 2000; // Prevent overwhelming corpus
const MAX_FILE_SIZE = 500000; // 500KB max per file
const MIN_FILE_SIZE = 50; // Skip tiny files

// Repository configurations
const REPOSITORIES = [
    {
        name: 'django/django',
        url: 'https://github.com/django/django.git',
        branch: 'main',
        paths: ['django/'],
        priority: 1 // Highest priority - most entries
    },
    {
        name: 'sympy/sympy', 
        url: 'https://github.com/sympy/sympy.git',
        branch: 'master',
        paths: ['sympy/'],
        priority: 2
    },
    {
        name: 'sphinx-doc/sphinx',
        url: 'https://github.com/sphinx-doc/sphinx.git', 
        branch: 'master',
        paths: ['sphinx/'],
        priority: 3
    },
    {
        name: 'matplotlib/matplotlib',
        url: 'https://github.com/matplotlib/matplotlib.git',
        branch: 'main', 
        paths: ['lib/matplotlib/'],
        priority: 4
    },
    {
        name: 'scikit-learn/scikit-learn',
        url: 'https://github.com/scikit-learn/scikit-learn.git',
        branch: 'main',
        paths: ['sklearn/'],
        priority: 5
    },
    {
        name: 'astropy/astropy',
        url: 'https://github.com/astropy/astropy.git',
        branch: 'main',
        paths: ['astropy/'],
        priority: 6
    },
    {
        name: 'pydata/xarray',
        url: 'https://github.com/pydata/xarray.git',
        branch: 'main', 
        paths: ['xarray/'],
        priority: 7
    },
    {
        name: 'pytest-dev/pytest',
        url: 'https://github.com/pytest-dev/pytest.git',
        branch: 'main',
        paths: ['src/pytest/', 'src/_pytest/'],
        priority: 8
    },
    {
        name: 'pylint-dev/pylint',
        url: 'https://github.com/pylint-dev/pylint.git',
        branch: 'main',
        paths: ['pylint/'],
        priority: 9
    },
    {
        name: 'psf/requests',
        url: 'https://github.com/psf/requests.git',
        branch: 'main',
        paths: ['src/requests/', 'requests/'],
        priority: 10
    }
];

function shouldSkipFile(filePath, stats) {
    // Skip if too large or too small
    if (stats.size > MAX_FILE_SIZE || stats.size < MIN_FILE_SIZE) {
        return true;
    }
    
    // Skip build artifacts and cache files
    const skipPatterns = [
        /\/__pycache__\//,
        /\.pyc$/,
        /\.pyo$/,
        /\.egg-info\//,
        /\/build\//,
        /\/dist\//,
        /\/\.git\//,
        /\/node_modules\//,
        /\/venv\//,
        /\/\.pytest_cache\//,
        /\/\.tox\//,
        /\/docs\/build\//,
        /\/htmlcov\//,
        /\.coverage$/,
        /\.DS_Store$/
    ];
    
    return skipPatterns.some(pattern => pattern.test(filePath));
}

function isPythonFile(filePath) {
    return filePath.endsWith('.py');
}

function cloneOrUpdateRepository(repo) {
    const repoPath = path.join(REPOS_DIR, repo.name.replace('/', '_'));
    
    try {
        if (fs.existsSync(repoPath)) {
            console.log(`📥 Updating ${repo.name}...`);
            execSync(`cd "${repoPath}" && git fetch origin && git checkout ${repo.branch} && git reset --hard origin/${repo.branch}`, 
                    { stdio: 'pipe' });
        } else {
            console.log(`🔄 Cloning ${repo.name}...`);
            fs.mkdirSync(path.dirname(repoPath), { recursive: true });
            execSync(`git clone --depth 1 --branch ${repo.branch} "${repo.url}" "${repoPath}"`, 
                    { stdio: 'pipe' });
        }
        return repoPath;
    } catch (error) {
        console.error(`❌ Failed to clone/update ${repo.name}:`, error.message);
        return null;
    }
}

function extractFilesFromPath(repoPath, sourcePath, targetPrefix, maxFiles) {
    let fileCount = 0;
    const fullSourcePath = path.join(repoPath, sourcePath);
    
    if (!fs.existsSync(fullSourcePath)) {
        console.warn(`⚠️ Path not found: ${fullSourcePath}`);
        return 0;
    }
    
    function walkDirectory(dir, relativePath = '') {
        if (fileCount >= maxFiles) return;
        
        try {
            const entries = fs.readdirSync(dir);
            
            for (const entry of entries) {
                if (fileCount >= maxFiles) break;
                
                const fullPath = path.join(dir, entry);
                const relativeFilePath = path.join(relativePath, entry);
                
                try {
                    const stats = fs.statSync(fullPath);
                    
                    if (shouldSkipFile(fullPath, stats)) {
                        continue;
                    }
                    
                    if (stats.isDirectory()) {
                        walkDirectory(fullPath, relativeFilePath);
                    } else if (stats.isFile() && isPythonFile(fullPath)) {
                        // Create flattened filename
                        const flattenedName = `${targetPrefix}_${relativeFilePath.replace(/[\/\\]/g, '_')}`;
                        const targetPath = path.join(CORPUS_DIR, flattenedName);
                        
                        // Copy file to corpus
                        try {
                            const content = fs.readFileSync(fullPath, 'utf8');
                            fs.writeFileSync(targetPath, content);
                            fileCount++;
                            
                            if (fileCount % 100 === 0) {
                                console.log(`   📁 ${fileCount} files extracted...`);
                            }
                        } catch (copyError) {
                            console.warn(`⚠️ Failed to copy ${fullPath}:`, copyError.message);
                        }
                    }
                } catch (statError) {
                    console.warn(`⚠️ Failed to stat ${fullPath}:`, statError.message);
                }
            }
        } catch (readError) {
            console.warn(`⚠️ Failed to read directory ${dir}:`, readError.message);
        }
    }
    
    walkDirectory(fullSourcePath);
    return fileCount;
}

function extractRepositoryFiles(repo) {
    const repoPath = cloneOrUpdateRepository(repo);
    if (!repoPath) {
        return 0;
    }
    
    console.log(`🔍 Extracting files from ${repo.name}...`);
    
    let totalFiles = 0;
    const filesPerPath = Math.floor(MAX_FILES_PER_REPO / repo.paths.length);
    
    for (const sourcePath of repo.paths) {
        const targetPrefix = `${repo.name.replace('/', '_')}_${sourcePath.replace(/[\/\\]/g, '_')}`.replace(/_+$/, '');
        const fileCount = extractFilesFromPath(repoPath, sourcePath, targetPrefix, filesPerPath);
        totalFiles += fileCount;
        console.log(`   ✅ ${fileCount} files from ${sourcePath}`);
    }
    
    return totalFiles;
}

async function main() {
    console.log('🚀 SWE-Bench Corpus Extraction Starting...');
    console.log('');
    
    // Ensure directories exist
    fs.mkdirSync(REPOS_DIR, { recursive: true });
    fs.mkdirSync(CORPUS_DIR, { recursive: true });
    
    // Clear existing corpus
    console.log('🧹 Clearing existing corpus...');
    const existingFiles = fs.readdirSync(CORPUS_DIR);
    for (const file of existingFiles) {
        if (file.endsWith('.py')) {
            fs.unlinkSync(path.join(CORPUS_DIR, file));
        }
    }
    
    let totalFiles = 0;
    const results = {};
    
    // Sort repositories by priority
    const sortedRepos = REPOSITORIES.sort((a, b) => a.priority - b.priority);
    
    for (const repo of sortedRepos) {
        console.log(`\n📦 Processing ${repo.name}...`);
        const fileCount = extractRepositoryFiles(repo);
        results[repo.name] = fileCount;
        totalFiles += fileCount;
        console.log(`✅ ${repo.name}: ${fileCount} files extracted`);
    }
    
    // Generate summary
    console.log('\n📊 EXTRACTION SUMMARY');
    console.log('======================');
    for (const [repo, count] of Object.entries(results)) {
        console.log(`${repo.padEnd(30)} ${count.toString().padStart(6)} files`);
    }
    console.log('======================');
    console.log(`${'TOTAL'.padEnd(30)} ${totalFiles.toString().padStart(6)} files`);
    
    // Generate extraction metadata
    const metadata = {
        extraction_time: new Date().toISOString(),
        total_files: totalFiles,
        repositories: results,
        extraction_config: {
            max_files_per_repo: MAX_FILES_PER_REPO,
            max_file_size: MAX_FILE_SIZE,
            min_file_size: MIN_FILE_SIZE
        }
    };
    
    fs.writeFileSync(path.join(CORPUS_DIR, 'extraction_metadata.json'), 
                     JSON.stringify(metadata, null, 2));
    
    console.log('\n🎉 SWE-Bench corpus extraction completed!');
    console.log(`📁 ${totalFiles} Python files extracted to ${CORPUS_DIR}/`);
    console.log(`📋 Metadata saved to ${CORPUS_DIR}/extraction_metadata.json`);
    
    if (totalFiles === 0) {
        console.error('\n❌ No files were extracted! Check repository access and paths.');
        process.exit(1);
    }
}

main().catch(error => {
    console.error('💥 Fatal error:', error);
    process.exit(1);
});